### PR TITLE
IRawValue + dbNullLayoutRenderer

### LINKED ISSUE #1332: [Feature Request] Support for something like log4net IRawLayout for targets that need objects.
- Why do we need it?
  I'm currently working to create an NLog target for a logging system that takes objects in a specific structure.  Many of the properties on the structure are non string (e.g. int, DateTime, etc.), and right now if I would like to take advantage of the Layout renderers to provide a bit more flexibility to users from the configuration I will need to take the string produced by the renderer and parse it back into the appropriate type.

In many cases such as the MDC or MDLC, the structure has all the properties I need in the proper types, so I can just manually lookup the properties myself, but if the MDC layout renderer supported a way to get the raw object version of a property I could skip the costly parsing and also my code won't need to checking in many different locations for values.

```
public interface IRawLayout
{
    public object RenderRaw(LogEventInfo logEvent);
}
```

Anything that supports IRawLayout should essentially be able to fall back to ILayout by just ToStringing the object, but we could also derive IRawLayout from ILayout to force the implementers to ensure they provide for the lowest common denominator.

I haven't looked TOO deeply into how this would work between the Layout/Renderer layer, so I'm not sure how easy it would be, but wanted to start a conversation to find out if this is something that has been discussed already.

log4net `IRawLayout`: https://svn.apache.org/viewvc/logging/log4net/trunk/src/Layout/IRawLayout.cs?view=markup

### PR DESCRIPTION
**Pull request by 304NotModified**
needed for #522 (see also #1435 or #2767)

Not sure if this is enough. How to log ${level} as int to the database? We like to prevent toString and int.parse.

options: 

1. database target aware of LogLevel type
2. or make IRawValue generic, e.g. IRawValue<LogLevel>, IRawValue<int>
3. or request type to GetRawValue, e.g. GetRawValue(typeof(int))  - could be optional.

todo decide

- [x] naming of getRawValue and IRawValue
- [x] also work with TypeCode? (provided or request), really useful is TypeCode.DBNull for the database target, see also https://github.com/NLog/NLog/issues/293
- [x] interface (IRawValue) or abstract class? Later for performance reasons. (also possible, virtual method on layout class, but could be tricky with null)

fixes #1332

todo 
 
- [x] impl for all needed layout renderers 
- [x] unit tests

### CODE REVIEW

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 1]
```diff
@@ -178,6 +178,11 @@ public List<ExceptionRenderingFormat> InnerFormats
             private set;
         }
 
+        /// <summary>
+        /// Get raw value, updates the sequence
+        /// </summary>
+        object IRawValue.GetRawValue(LogEventInfo logEvent) => logEvent.Exception;
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: CRITICAL Rename parameter 'logEvent' to 'logEventInfo' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

File: src/NLog/LayoutRenderers/GuidLayoutRenderer.cs [hunk 2]
```diff
@@ -67,13 +67,26 @@ public GuidLayoutRenderer()
         [DefaultValue(false)]
         public bool GeneratedFromLogEvent { get; set; }
 
+        /// <summary>
+        /// Get raw value
+        /// </summary>
+        object IRawValue.GetRawValue(LogEventInfo logEvent) => GetValue(logEvent);
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/GuidLayoutRenderer.cs]: CRITICAL Rename parameter 'logEvent' to 'logEventInfo' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

File: src/NLog/LayoutRenderers/CounterLayoutRenderer.cs [hunk 3]
```diff
@@ -82,6 +82,18 @@ public CounterLayoutRenderer()
...
+        {
+            int v = GetNextValue(logEvent);
+
+            builder.AppendInvariant(v);
+        }
+
+        /// <summary>
+        /// Get raw value, updates the sequence
+        /// </summary>
+        object IRawValue.GetRawValue(LogEventInfo logEvent) => GetNextValue(logEvent);
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/CounterLayoutRenderer.cs]: CRITICAL Rename parameter 'logEvent' to 'logEventInfo' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

File: src/NLog/LayoutRenderers/GdcLayoutRenderer.cs [hunk 4]
```diff
@@ -66,10 +66,20 @@ public class GdcLayoutRenderer : LayoutRenderer
...
-            var value = GlobalDiagnosticsContext.GetObject(Item);
+            object value = GetValue();
             var formatProvider = GetFormatProvider(logEvent, null);
             builder.AppendFormattedValue(value, Format, formatProvider);
         }
+
+        /// <summary>
+        /// Get raw value.
+        /// </summary>
+        object IRawValue.GetRawValue(LogEventInfo logEvent) => GetValue();
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/GdcLayoutRenderer.cs]: CRITICAL Rename parameter 'logEvent' to 'logEventInfo' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

File: src/NLog/LayoutRenderers/EventPropertiesLayoutRenderer.cs [hunk 5]
```diff
@@ -83,12 +83,26 @@ public EventPropertiesLayoutRenderer()
...
+        private bool GetValue(LogEventInfo logEvent, out object value)
+        {
+            value = null;
+            return logEvent.HasProperties && logEvent.Properties.TryGetValue(Item, out value);
+        }
+
+        /// <summary>
+        /// Get raw value, updates the sequence
+        /// </summary>
+        object IRawValue.GetRawValue(LogEventInfo logEvent)
```
Comment (NLog-SonarQube) [inline src/NLog/LayoutRenderers/EventPropertiesLayoutRenderer.cs]: CRITICAL Rename parameter 'logEvent' to 'logEventInfo' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

**Comment by NLog-SonarQube** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 5 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

**Comment by 304NotModified** [comment]
@snakefoot like your opinion on this :)

**Comment by snakefoot** [comment]
@304NotModified Like the RawValue idea. Makes good sense when it comes to converting into parameter-types.

Regarding conversion of a RawValue to other types (Ex. LogLevel to Ordinal) then I guess one could reuse the type-invariance that has been made for when-filtering.

File: src/NLog/Layouts/SimpleLayout.cs [hunk 6]
```diff
@@ -240,6 +240,18 @@ internal void SetRenderers(LayoutRenderer[] renderers, string text)
...
+        /// <inheritdoc />
+        public override object GetRawValue(LogEventInfo logEvent)
+        {
+
+            if (Renderers.Count == 1 && Renderers[0] is IRawValue rawValueLayoutRenderer)
+            {
+                return rawValueLayoutRenderer.GetRawValue(logEvent);
+            }
+
+            return null;
```
Comment (304NotModified) [inline src/NLog/Layouts/SimpleLayout.cs]: doubting if this is good enough..

If we get `null`, then we could call render again and that's not always the best solution

Comment (snakefoot) [reply src/NLog/Layouts/SimpleLayout.cs]: I think it is a good idea that `Layout.GetRawValue` always return null (Good default for CsvLayout + JsonLayout + etc.). Then SimpleLayout need to do the conversion if ``IRawValue`-optimization is not available.

Comment (snakefoot) [reply src/NLog/Layouts/SimpleLayout.cs]: Prefer that the fallback render logic is inside this method, instead of the caller having to handle `null`.

Comment (304NotModified) [reply src/NLog/Layouts/SimpleLayout.cs]: I changed it already to TryGetRawValue, for performance reasons.

If it returns null (out) and true, there is no need to use the string render anymore

Comment (snakefoot) [reply src/NLog/Layouts/SimpleLayout.cs]: Prefer the old logic, where it tries all kinds of ways, and then fallback to a provided default value.

Comment (304NotModified) [reply src/NLog/Layouts/SimpleLayout.cs]: but then I could call twice .render, while that's not needed

Comment (snakefoot) [reply src/NLog/Layouts/SimpleLayout.cs]: You are right `TryGetRawValue` will just be a single helper method, that is used by an generic-extension method that gets the default-value as input paramater and can perform rendering if needed.

File: src/NLog/LayoutRenderers/IdentityLayoutRenderer.cs [hunk 7]
```diff
@@ -95,36 +95,45 @@ public IdentityLayoutRenderer()
...
-                    }
+                    builder.Append(separator);
+                    builder.Append(identity.Name);
                 }
             }
+
+        }
+
+        /// <inheritdoc />
+        object IRawValue.GetRawValue(LogEventInfo logEvent)
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/IdentityLayoutRenderer.cs]: @snakefoot do you think this layout renderer should implement IRawValue, and if so, should use the options IsAuthenticated/AuthType/Name is the rawValue? (well then it's a string)

Comment (304NotModified) [reply src/NLog/LayoutRenderers/IdentityLayoutRenderer.cs]: OK thanks! done

**Comment by 304NotModified** [comment]
some docs here: https://github.com/NLog/NLog/wiki/Database-target

**Pull request merged**